src/main/java/org/zalando/nakadi/validation/SchemaCompatibilityChecker.java

```diff
@@ -0,0 +1,114 @@
+package org.zalando.nakadi.validation;
+
+import org.everit.json.schema.ArraySchema;
+import org.everit.json.schema.CombinedSchema;
+import org.everit.json.schema.ObjectSchema;
+import org.everit.json.schema.ReferenceSchema;
+import org.everit.json.schema.Schema;
+import org.zalando.nakadi.validation.schema.SchemaConstraint;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Stack;
+
+public class SchemaCompatibilityChecker {
+
+    private final List<SchemaConstraint> constraints;
+
+    public SchemaCompatibilityChecker(final List<SchemaConstraint> constraints) {
+        this.constraints = constraints;
+    }
+
+    public List<SchemaIncompatibility> checkConstraints(final Schema schema) {
+        final List<SchemaIncompatibility> incompatibilities = new ArrayList<>();
+
+        recursiveCheckConstraints(schema, new Stack<>(), incompatibilities);
+
+        return incompatibilities;
+    }
+
+    private void recursiveCheckConstraints(
+            final Schema schema,
+            final Stack<String> jsonPath,
+            final List<SchemaIncompatibility> schemaIncompatibilities) {
+
+        for (final SchemaConstraint constraint : constraints) {
+            final Optional<SchemaIncompatibility> incompatibility = constraint.validate(jsonPath, schema);
+            if (incompatibility.isPresent()) {
+                schemaIncompatibilities.add(incompatibility.get());
+            }
+        }
+
+        if (schema instanceof ObjectSchema) {
```

Comment:
It's not. We do not own "Schema" class since it's from a third library. We could wrap then to some Nakadi owned class, but I considered it would be too much indirection for a small gain.